pandas/io/formats/format.py

```diff
@@ -879,14 +879,23 @@ def format_array(values, formatter, float_format=None, na_rep='NaN',
     List[str]
     """
 
+    if is_extension_array_dtype(values.dtype):
+        if isinstance(values, (ABCIndexClass, ABCSeries)):
+            values = values._values
+
+        if is_categorical_dtype(values.dtype):
```

Comment:
why would we not have both of these in the below if/elif clause?

Reply:
we're deciding which values to use here in the same way as ExtensionArrayFormatter did.

the following if/else clause is selecting the Formatter class to use based on those values.

Reply:
Seems like the idea is to extract the datetime64[ns] from the Categorical, and then reuse the Datetime64Formatter by going into the if / elif below.

Reply:
yes i get the idea, trying to see if the logic can somehow be simpler

Reply:
We could potentially define `Categorical._formatter`, which would provide the appropriate scalar formatter based on it's `.dtype`?  Not sure if that'll work or not.

Reply:
hmm we actually define this
```
    def _formatter(self, boxed=False):
        # Defer to CategoricalFormatter's formatter.
        return None
```

Reply:
maybe we are not really using this attribute fully?

Reply:
format_array is used by `to_string`, `to_html`, `_repr_html`, `to_latex` and for the `repr` of many objects. 

it should in theory be simpler and more generic.

it may be beneficial to move some logic out into the objects themselves so that `format_array` can work with any extension array and not require this special casing. (i think that is outside of the scope of this PR)

This PR is intended to remove the call to `format_array` from within `ExtensionArrayFormatter` so that the `formatter` parameter of `format_array` can be used for custom formatters wihout defaults being applied.

`ExtensionArrayFormatter` was dispatching back to `format_array` to then dispatch to the appropriate (another) Formatter. 


> maybe we are not really using this attribute fully?

agreed. many just return None to defer to the Formatters.